### PR TITLE
R3

### PR DESCRIPTION
**Pull request by santhh**
Please review this release for merge. 
- Efficient use of split DoFn for large files. It's no longer uses buffered reader to access file between line ranges.
- Only one call to KMS apis if needed /bundle for customer supplied keys
- Dynamic file generation and BQ tables / window. 
-  Upgrade to Beam 2.8 and latest DLP Java SDKs.
- Performance tested for 140 MB file (150M rows). Used 500 DLP quotas with 5 highmem-8 type machines took less than a minute to complete.

### CODE REVIEW

File: src/main/java/com/google/swarm/tokenization/common/CSVReader.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+				e.printStackTrace();
+			} catch (GeneralSecurityException e) {
+
+				e.printStackTrace();
+			}
+		}
+		String bucketName = Util.parseBucketName(f.getMetadata().resourceId().getCurrentDirectory().toString());
+
+		String objectName = f.getMetadata().resourceId().getFilename().toString();
+		String[] fileKey = objectName.split("\\.", 2);
```
Comment (theodoresiu) [inline src/main/java/com/google/swarm/tokenization/common/CSVReader.java]: Does splitting on `"\\." `work for getting the fileKey?

File: src/main/java/com/google/swarm/tokenization/common/DLPTokenizationDoFn.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+					.setDeidentifyTemplateName(this.deIdentifyTemplateName.get()).setItem(tableItem).build();
+
+		}
+
+		response = dlpServiceClient.deidentifyContent(request);
+		Table encryptedData = response.getItem().getTable();
+		LOG.info("Request Size Successfully Tokenized:{} rows {} bytes ", encryptedData.getRowsList().size(),
+				request.toByteString().size());
+
+		List<String> outputHeaders = encryptedData.getHeadersList().stream().map(FieldId::getName)
```
Comment (theodoresiu) [inline src/main/java/com/google/swarm/tokenization/common/DLPTokenizationDoFn.java]: Should we put this `get headers list/check` in a method?

File: src/main/java/com/google/swarm/tokenization/common/Util.java [hunk 3]
```diff
@@ -133,14 +115,22 @@ public static BufferedReader getReader(boolean customerSuppliedKey, String objec
 
 	}
 
+	public static String checkHeaderName(String name) {
```
Comment (theodoresiu) [inline src/main/java/com/google/swarm/tokenization/common/Util.java]: Yes, this was what I referring to in our previous talks. Bigquery does not allow spaces and other special characters. Should we expand the checker to remove all special characters?

Comment (theodoresiu) [reply src/main/java/com/google/swarm/tokenization/common/Util.java]: See below for a suggestion

File: src/main/java/com/google/swarm/tokenization/common/Util.java [hunk 4]
```diff
@@ -133,14 +115,22 @@ public static BufferedReader getReader(boolean customerSuppliedKey, String objec
...
+		String checkedHeader = name.replaceAll("\\s", "");
+		checkedHeader = checkedHeader.replaceAll("'", "");
+		checkedHeader = checkedHeader.replaceAll("/", "");
```
Comment (theodoresiu) [inline src/main/java/com/google/swarm/tokenization/common/Util.java]:
```suggestion
		checkedHeader = checkedHeader.replaceAll("\\W", "");
```

File: src/main/java/com/google/swarm/tokenization/common/Util.java [hunk 5]
```diff
@@ -133,14 +115,22 @@ public static BufferedReader getReader(boolean customerSuppliedKey, String objec
 
 	}
 
+	public static String checkHeaderName(String name) {
+		String checkedHeader = name.replaceAll("\\s", "");
```
Comment (theodoresiu) [inline src/main/java/com/google/swarm/tokenization/common/Util.java]:
```suggestion
		String checkedHeader = name.replaceAll("\\s", "_");
```

File: build.gradle [hunk 6]
```diff
@@ -59,11 +66,12 @@ dependencies {
...
+    compile 'com.google.cloud:google-cloud-dlp:0.74.0-beta'
     compile 'com.google.guava:guava:27.0-jre'
     compile 'com.google.api-client:google-api-client:1.27.0'
     compile group: 'com.google.apis', name: 'google-api-services-cloudkms', version: 'v1-rev53-1.23.0'
-
+	compile "com.google.auto.value:auto-value-annotations:1.6.2"
+    annotationProcessor "com.google.auto.value:auto-value:1.6.2"
 	testCompile group: 'org.apache.beam', name: 'beam-runners-direct-java', version: dataflowBeamVersion
 	testCompile group: 'org.slf4j', name: 'slf4j-jdk14', version: '1.7.5'
 	testCompile 'junit:junit:4.12'
```
[the remark this answers is not in the document]
Comment (santhh) [reply build.gradle]: @theodoresiu This is to handle dynamic class loader passes in to gradle run. We have two pipelines and need to one which one to execute in runtime. For the local build and test, you can either setup the param or comment out this line for the specific class def.

File: dlp-tokenization_metadata [hunk 7]
```diff
@@ -16,28 +16,23 @@
 	 "is_optional": false
   },
    {
-     "name": "tableSpec",
-     "label": "BQ Table Spec ",
-     "help_text": "<project_id>:<dataset_id>.<table_id>",
+     "name": "dataset",
+     "label": "BQ DatasetId e.g pii_dataset ",
+     "help_text": "<dataset_id>",
```
[the remark this answers is not in the document]
Comment (santhh) [reply dlp-tokenization_metadata]: @theodoresiu Yes there are some validation need to be done for the metadata template file. I will create a FR for next release.

File: src/main/java/com/google/swarm/tokenization/common/DLPTokenizationDoFn.java (hunk 2 above)
Comment (santhh) [reply src/main/java/com/google/swarm/tokenization/common/DLPTokenizationDoFn.java]: @theodoresiu getHeaderList is part of DLP table API. Do you mean add some validation to it before joining?

File: src/main/java/com/google/swarm/tokenization/common/Util.java (hunk 3 above)
Comment (santhh) [reply src/main/java/com/google/swarm/tokenization/common/Util.java]: @theodoresiu  yes!  Can you create a PR after i merge?

File: src/main/java/com/google/swarm/tokenization/common/CSVReader.java (hunk 1 above)
Comment (santhh) [reply src/main/java/com/google/swarm/tokenization/common/CSVReader.java]: Yes. Did you find any issue?

**Pull request merged**